File: packages/lynx-webrtc/ios/WebRTCModule.swift
import AVFoundation
import Foundation
import Lynx
import WebRTC

/// Native WebRTC bridge — JS-callable side.
///
/// JS usage (via the `@sigx/lynx-webrtc` shim, not directly):
///
///   NativeModules.WebRTC.createPeer(id, config, cb)
///   NativeModules.WebRTC.createOffer(id, options, cb)
///   NativeModules.WebRTC.getUserMedia(trackId, constraints, cb)
///   ...
///
/// State lives in `PeerConnectionStore`. Async events (state changes, ICE
/// candidates, remote tracks, data channel traffic) flow through
/// `WebRTCEventBus` → per-LynxView `WebRTCPublisher` → JS
/// `GlobalEventEmitter`. Microphone permission goes through
/// `AVAudioSession.recordPermission`, like `@sigx/lynx-audio`.
@objc class WebRTCModule: NSObject, LynxModule {

    @objc static var name: String { "WebRTC" }

    @objc static var methodLookup: [String: String] {
        [
            "getUserMedia":         NSStringFromSelector(#selector(getUserMedia(_:constraints:callback:))),
            "requestPermission":    NSStringFromSelector(#selector(requestPermission(_:))),
            "getPermissionStatus":  NSStringFromSelector(#selector(getPermissionStatus(_:))),
            "createPeer":           NSStringFromSelector(#selector(createPeer(_:config:callback:))),
            "createOffer":          NSStringFromSelector(#selector(createOffer(_:options:callback:))),
            "createAnswer":         NSStringFromSelector(#selector(createAnswer(_:options:callback:))),
            "setLocalDescription":  NSStringFromSelector(#selector(setLocalDescription(_:desc:callback:))),
            "setRemoteDescription": NSStringFromSelector(#selector(setRemoteDescription(_:desc:callback:))),
            "addIceCandidate":      NSStringFromSelector(#selector(addIceCandidate(_:candidate:callback:))),
            "addTrack":             NSStringFromSelector(#selector(addTrack(_:trackId:streamIds:callback:))),
            "removeTrack":          NSStringFromSelector(#selector(removeTrack(_:senderId:callback:))),
            "createDataChannel":    NSStringFromSelector(#selector(createDataChannel(_:dcId:label:initDict:callback:))),
            "dcSend":               NSStringFromSelector(#selector(dcSend(_:payload:isBinary:callback:))),
            "dcClose":              NSStringFromSelector(#selector(dcClose(_:callback:))),
            "setTrackEnabled":      NSStringFromSelector(#selector(setTrackEnabled(_:enabled:callback:))),
            "stopTrack":            NSStringFromSelector(#selector(stopTrack(_:callback:))),
            "closePeer":            NSStringFromSelector(#selector(closePeer(_:callback:))),
            "setAudioOutput":       NSStringFromSelector(#selector(setAudioOutput(_:callback:))),
        ]
    }

    required override init() { super.init() }
    required init(param: Any) { super.init() }

    // MARK: - Media

    // `constraints` keeps the JS-callable selector; v1 captures with platform defaults.
    @objc func getUserMedia(_ trackId: NSNumber, constraints _: [String: Any]?, callback: LynxCallbackBlock?) {
        // Browser semantics: prompt whenever the permission is promptable,
        // resolve immediately when granted, fail fast when denied (iOS never
        // re-prompts — the user must visit Settings).
        let session = AVAudioSession.sharedInstance()
        switch session.recordPermission {
        case .granted:
            callback?(Self.captureResult(trackId: trackId.intValue))
        case .undetermined:
            session.requestRecordPermission { granted in
                if granted {
                    callback?(Self.captureResult(trackId: trackId.intValue))
                } else {
                    callback?(Self.deniedResult())
                }
            }
        default:
            callback?(Self.deniedResult())
        }
    }

    private static func captureResult(trackId: Int) -> [String: Any] {
        if let error = PeerConnectionStore.shared.createMicrophoneTrack(trackId: trackId) {
            return ["error": error]
        }
        return ["label": "microphone"]
    }

    private static func deniedResult() -> [String: Any] {
        return ["error": "Microphone permission not granted", "errorName": "NotAllowedError"]
    }

    @objc func setTrackEnabled(_ trackId: NSNumber, enabled: NSNumber, callback: LynxCallbackBlock?) {
        callback?(Self.result(PeerConnectionStore.shared.setTrackEnabled(
            trackId: trackId.intValue, enabled: enabled.boolValue
        )))
    }

    @objc func stopTrack(_ trackId: NSNumber, callback: LynxCallbackBlock?) {
        callback?(Self.result(PeerConnectionStore.shared.stopTrack(trackId: trackId.intValue)))
    }

    // MARK: - Peer connection

    @objc func createPeer(_ peerId: NSNumber, config: [String: Any]?, callback: LynxCallbackBlock?) {
        callback?(Self.result(PeerConnectionStore.shared.createPeer(peerId: peerId.intValue, config: config)))
    }

    @objc func createOffer(_ peerId: NSNumber, options: [String: Any]?, callback: LynxCallbackBlock?) {
        PeerConnectionStore.shared.createOffer(peerId: peerId.intValue, options: options) { sdp, error in
            callback?(Self.descriptionResult(sdp, error))
        }
    }

    // `options` keeps signature parity with createOffer; answers take none.
    @objc func createAnswer(_ peerId: NSNumber, options _: [String: Any]?, callback: LynxCallbackBlock?) {
        PeerConnectionStore.shared.createAnswer(peerId: peerId.intValue) { sdp, error in
            callback?(Self.descriptionResult(sdp, error))
        }
    }

    @objc func setLocalDescription(_ peerId: NSNumber, desc: [String: Any]?, callback: LynxCallbackBlock?) {
        PeerConnectionStore.shared.setLocalDescription(peerId: peerId.intValue, desc: desc) { applied, error in
            callback?(Self.descriptionResult(applied, error))
        }
    }

    @objc func setRemoteDescription(_ peerId: NSNumber, desc: [String: Any]?, callback: LynxCallbackBlock?) {
        PeerConnectionStore.shared.setRemoteDescription(peerId: peerId.intValue, desc: desc) { error in
            callback?(Self.result(error))
        }
    }

    @objc func addIceCandidate(_ peerId: NSNumber, candidate: [String: Any]?, callback: LynxCallbackBlock?) {
        PeerConnectionStore.shared.addIceCandidate(peerId: peerId.intValue, candidate: candidate) { error in
            callback?(Self.result(error))
        }
    }

    @objc func addTrack(_ peerId: NSNumber, trackId: NSNumber, streamIds: [Any]?, callback: LynxCallbackBlock?) {
        let ids = (streamIds as? [String])?.filter { !$0.isEmpty } ?? []
        let (senderId, error) = PeerConnectionStore.shared.addTrack(
            peerId: peerId.intValue, trackId: trackId.intValue, streamIds: ids
        )
        if let error = error {
            callback?(["error": error])
            return
        }
        callback?(["senderId": senderId])
    }

    @objc func removeTrack(_ peerId: NSNumber, senderId: NSNumber, callback: LynxCallbackBlock?) {
        callback?(Self.result(PeerConnectionStore.shared.removeTrack(
            peerId: peerId.intValue, senderId: senderId.intValue
        )))
    }

    @objc func closePeer(_ peerId: NSNumber, callback: LynxCallbackBlock?) {
        callback?(Self.result(PeerConnectionStore.shared.closePeer(peerId: peerId.intValue)))
    }

    // MARK: - Data channels

    @objc func createDataChannel(_ peerId: NSNumber, dcId: NSNumber, label: String?, initDict: [String: Any]?, callback: LynxCallbackBlock?) {
        callback?(Self.result(PeerConnectionStore.shared.createDataChannel(
            peerId: peerId.intValue, dcId: dcId.intValue, label: label ?? "", initDict: initDict
        )))
    }

    @objc func dcSend(_ dcId: NSNumber, payload: String?, isBinary: NSNumber?, callback: LynxCallbackBlock?) {
        callback?(Self.result(PeerConnectionStore.shared.dcSend(
            dcId: dcId.intValue, payload: payload ?? "", isBinary: isBinary?.boolValue ?? false
        )))
    }

    @objc func dcClose(_ dcId: NSNumber, callback: LynxCallbackBlock?) {
        callback?(Self.result(PeerConnectionStore.shared.dcClose(dcId: dcId.intValue)))
    }

    // MARK: - Audio routing & permission

    @objc func setAudioOutput(_ route: String?, callback: LynxCallbackBlock?) {
        callback?(Self.result(WebRTCAudioController.shared.setRoute(route ?? "")))
    }

    @objc func requestPermission(_ callback: LynxCallbackBlock?) {
        let session = AVAudioSession.sharedInstance()
        switch session.recordPermission {
        case .granted:
            callback?(["status": "granted", "canAskAgain": false])
        case .denied:
            // iOS won't re-prompt on second call — user must visit Settings.
            callback?(["status": "denied", "canAskAgain": false])
        case .undetermined:
            session.requestRecordPermission { granted in
                callback?([
                    "status": granted ? "granted" : "denied",
                    "canAskAgain": false,
                ])
            }
        @unknown default:
            callback?(["status": "denied", "canAskAgain": false])
        }
    }

    @objc func getPermissionStatus(_ callback: LynxCallbackBlock?) {
        switch AVAudioSession.sharedInstance().recordPermission {
        case .granted:
            callback?(["status": "granted", "canAskAgain": false])
        case .denied:
            callback?(["status": "denied", "canAskAgain": false])
        case .undetermined:
            callback?(["status": "undetermined", "canAskAgain": true])
        @unknown default:
            callback?(["status": "denied", "canAskAgain": false])
        }
    }

    // MARK: - Helpers

    private static func result(_ error: String?) -> Any {
        if let error = error { return ["error": error] }
        return [String: Any]()
    }

    private static func descriptionResult(_ sdp: RTCSessionDescription?, _ error: String?) -> Any {
        if let error = error { return ["error": error] }
        guard let sdp = sdp else { return [String: Any]() }
        return ["type": PeerConnectionStore.sdpTypeString(sdp.type), "sdp": sdp.sdp]
    }
}

File: packages/lynx-webrtc/android/com/sigx/webrtc/WebRTCModule.kt
package com.sigx.webrtc

import android.content.Context
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.JavaOnlyMap
import com.lynx.react.bridge.ReadableArray
import com.lynx.react.bridge.ReadableMap
import com.sigx.permissions.PermissionHelper
import org.webrtc.SessionDescription

/**
 * Native WebRTC bridge — JS-callable side.
 *
 * JS usage (via the `@sigx/lynx-webrtc` shim, not directly):
 *
 *   NativeModules.WebRTC.createPeer(id, config, cb)
 *   NativeModules.WebRTC.createOffer(id, options, cb)
 *   NativeModules.WebRTC.getUserMedia(trackId, constraints, cb)
 *   ...
 *
 * State lives in [PeerConnectionStore]. Async events (state changes, ICE
 * candidates, remote tracks, data channel traffic) flow through
 * [WebRTCEventBus] → per-LynxView [WebRTCPublisher] → JS
 * `GlobalEventEmitter`. Microphone permission is delegated to
 * `@sigx/lynx-permissions`' `PermissionHelper` (`"microphone"` →
 * `RECORD_AUDIO`).
 */
class WebRTCModule(context: Context) : LynxModule(context) {

    // MARK: - Media

    // `constraints` is part of the JS-callable signature; v1 ignores it and
    // captures with platform defaults.
    @Suppress("UNUSED_PARAMETER")
    @LynxMethod
    fun getUserMedia(trackId: Int, constraints: ReadableMap?, callback: Callback?) {
        // Browser semantics: prompt whenever the permission is promptable,
        // resolve immediately when granted, fail fast only when blocked.
        when (PermissionHelper.checkPermission(mContext, "microphone").getString("status")) {
            "granted" -> callback?.invoke(captureResult(trackId))
            "blocked" -> callback?.invoke(deniedMap())
            else -> PermissionHelper.requestPermission(mContext, "microphone") { result ->
                if (result.getString("status") == "granted") {
                    callback?.invoke(captureResult(trackId))
                } else {
                    callback?.invoke(deniedMap())
                }
            }
        }
    }

    private fun captureResult(trackId: Int): JavaOnlyMap {
        val error = PeerConnectionStore.createMicrophoneTrack(mContext, trackId)
        if (error != null) return errorMap(error)
        return JavaOnlyMap().apply { putString("label", "microphone") }
    }

    @LynxMethod
    fun setTrackEnabled(trackId: Int, enabled: Boolean, callback: Callback?) {
        callback?.invoke(resultMap(PeerConnectionStore.setTrackEnabled(trackId, enabled)))
    }

    @LynxMethod
    fun stopTrack(trackId: Int, callback: Callback?) {
        callback?.invoke(resultMap(PeerConnectionStore.stopTrack(trackId)))
    }

    // MARK: - Peer connection

    @LynxMethod
    fun createPeer(peerId: Int, config: ReadableMap?, callback: Callback?) {
        callback?.invoke(resultMap(PeerConnectionStore.createPeer(mContext, peerId, config)))
    }

    @LynxMethod
    fun createOffer(peerId: Int, options: ReadableMap?, callback: Callback?) {
        PeerConnectionStore.createOffer(peerId, options) { desc, error ->
            callback?.invoke(descriptionMap(desc, error))
        }
    }

    // `options` keeps signature parity with createOffer; answers take none.
    @Suppress("UNUSED_PARAMETER")
    @LynxMethod
    fun createAnswer(peerId: Int, options: ReadableMap?, callback: Callback?) {
        PeerConnectionStore.createAnswer(peerId) { desc, error ->
            callback?.invoke(descriptionMap(desc, error))
        }
    }

    @LynxMethod
    fun setLocalDescription(peerId: Int, desc: ReadableMap?, callback: Callback?) {
        PeerConnectionStore.setLocalDescription(peerId, desc) { applied, error ->
            callback?.invoke(descriptionMap(applied, error))
        }
    }

    @LynxMethod
    fun setRemoteDescription(peerId: Int, desc: ReadableMap?, callback: Callback?) {
        PeerConnectionStore.setRemoteDescription(peerId, desc) { error ->
            callback?.invoke(resultMap(error))
        }
    }

    @LynxMethod
    fun addIceCandidate(peerId: Int, candidate: ReadableMap?, callback: Callback?) {
        PeerConnectionStore.addIceCandidate(peerId, candidate) { error ->
            callback?.invoke(resultMap(error))
        }
    }

    @LynxMethod
    fun addTrack(peerId: Int, trackId: Int, streamIds: ReadableArray?, callback: Callback?) {
        val ids = mutableListOf<String>()
        if (streamIds != null) {
            for (i in 0 until streamIds.size()) {
                streamIds.getString(i)?.let { if (it.isNotEmpty()) ids.add(it) }
            }
        }
        val (senderId, error) = PeerConnectionStore.addTrack(peerId, trackId, ids)
        if (error != null) {
            callback?.invoke(errorMap(error))
            return
        }
        callback?.invoke(JavaOnlyMap().apply { putInt("senderId", senderId) })
    }

    @LynxMethod
    fun removeTrack(peerId: Int, senderId: Int, callback: Callback?) {
        callback?.invoke(resultMap(PeerConnectionStore.removeTrack(peerId, senderId)))
    }

    @LynxMethod
    fun closePeer(peerId: Int, callback: Callback?) {
        callback?.invoke(resultMap(PeerConnectionStore.closePeer(mContext, peerId)))
    }

    // MARK: - Data channels

    @LynxMethod
    fun createDataChannel(peerId: Int, dcId: Int, label: String?, init: ReadableMap?, callback: Callback?) {
        callback?.invoke(
            resultMap(PeerConnectionStore.createDataChannel(peerId, dcId, label ?: "", init))
        )
    }

    @LynxMethod
    fun dcSend(dcId: Int, payload: String?, isBinary: Boolean, callback: Callback?) {
        callback?.invoke(resultMap(PeerConnectionStore.dcSend(dcId, payload ?: "", isBinary)))
    }

    @LynxMethod
    fun dcClose(dcId: Int, callback: Callback?) {
        callback?.invoke(resultMap(PeerConnectionStore.dcClose(dcId)))
    }

    // MARK: - Audio routing & permission

    @LynxMethod
    fun setAudioOutput(route: String?, callback: Callback?) {
        callback?.invoke(resultMap(WebRTCAudioController.setRoute(mContext, route ?: "")))
    }

    @LynxMethod
    fun requestPermission(callback: Callback?) {
        PermissionHelper.requestPermission(mContext, "microphone") { result ->
            callback?.invoke(result)
        }
    }

    @LynxMethod
    fun getPermissionStatus(callback: Callback?) {
        callback?.invoke(PermissionHelper.checkPermission(mContext, "microphone"))
    }

    // MARK: - Helpers

    private fun resultMap(error: String?): JavaOnlyMap =
        if (error == null) JavaOnlyMap() else errorMap(error)

    private fun descriptionMap(desc: SessionDescription?, error: String?): JavaOnlyMap {
        if (error != null) return errorMap(error)
        return JavaOnlyMap().apply {
            if (desc != null) {
                putString("type", desc.type.canonicalForm())
                putString("sdp", desc.description)
            }
        }
    }

    private fun deniedMap(): JavaOnlyMap = JavaOnlyMap().apply {
        putString("error", "Microphone permission not granted")
        putString("errorName", "NotAllowedError")
    }

    private fun errorMap(message: String): JavaOnlyMap =
        JavaOnlyMap().apply { putString("error", message) }
}
